p2p/net/udp.go

```diff
@@ -0,0 +1,121 @@
+package net
+
+import (
+	"net"
+
+	"github.com/spacemeshos/go-spacemesh/log"
+
+	"github.com/spacemeshos/go-spacemesh/p2p/config"
+	"github.com/spacemeshos/go-spacemesh/p2p/node"
+)
+
+const maxMessageSize = 2048
+
+type UDPMessageEvent struct {
+	FromAddr net.Addr
+	Message  []byte
+}
+
+type UDPNet struct {
+	local      *node.LocalNode
+	logger     log.Log
+	udpAddress *net.UDPAddr
+	config     config.Config
+	msgChan    chan UDPMessageEvent
+	conn       *net.UDPConn
+	shutdown   chan struct{}
+}
+
+func NewUDPNet(config config.Config, local *node.LocalNode, log log.Log) (*UDPNet, error) {
+	addr, err := net.ResolveUDPAddr("udp", local.Address())
+	if err != nil {
+		return nil, err
+	}
+
+	return &UDPNet{
+		local:      local,
+		logger:     log,
+		udpAddress: addr,
+		config:     config,
+		msgChan:    make(chan UDPMessageEvent, config.BufferSize),
+		shutdown:   make(chan struct{}),
+	}, nil
+}
+
+func (n *UDPNet) Start() error {
+	listener, err := newUDPListener(n.udpAddress)
+	if err != nil {
+		return err
+	}
+	n.conn = listener
+	go n.listenToUDPNetworkMessages(listener)
+
+	return nil
+}
+
+// Shutdown stops listening and closes the connection
+func (n *UDPNet) Shutdown() {
+	close(n.shutdown)
+	if n.conn != nil {
+		n.conn.Close()
+	}
+}
+
+func newUDPListener(listenAddress *net.UDPAddr) (*net.UDPConn, error) {
+	//todo: grab different udp port from config
+	listen, err := net.ListenUDP("udp", listenAddress)
+	if err != nil {
+		return nil, err
+	}
+	err = listen.SetReadBuffer(maxMessageSize)
+	if err != nil {
+		return nil, err
+	}
+	return listen, nil
+}
+
+func (n *UDPNet) Send(to node.Node, data []byte) error {
+	// todo : handle session if not exist
+	addr, err := net.ResolveUDPAddr("udp", to.Address())
```

Comment:
why is there a different resolving algorithm for an address? it's a mapping between nodeid -> ip so should it be the same one as in tcp?

Reply:
There's no difference. this method is called after we've resolved the IP address by the NodeID.
 it's just a resolve method of the IP itself. (so hostnames could be used for example). so this just makes the address string we have about this node to a valid golang UDP address struct.